src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -0,0 +1,295 @@
+        {
+            if (_readResult.Buffer.Length > _inputLength)
+            {
+                _readResult = new ReadResult(_readResult.Buffer.Slice(0, _inputLength), _readResult.IsCanceled, isCompleted: true);
+            }
+            else if (_readResult.Buffer.Length == _inputLength)
+            {
+                _readResult = new ReadResult(_readResult.Buffer, _readResult.IsCanceled, isCompleted: true);
+            }
+
+            if (_readResult.IsCompleted)
+            {
+                TryStop();
+            }
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            if (_inputLength == 0)
+            {
+                return;
+            }
+
+            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
+
+            _inputLength -= dataLength;
+
+            _context.Input.AdvanceTo(consumed, examined);
+
+            OnDataRead(dataLength);
+        }
+
+        protected override void OnReadStarting()
+        {
+            if (_contentLength > _context.MaxRequestBodySize)
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTooLarge);
+            }
+        }
+
+        private ValueTask<ReadResult> StartTimingReadAsync(CancellationToken cancellationToken)
+        {
+            var readAwaitable = _context.Input.ReadAsync(cancellationToken);
+
+            if (!readAwaitable.IsCompleted && _timingEnabled)
+            {
+                TryProduceContinue();
+
+                _backpressure = true;
+                _context.TimeoutControl.StartTimingRead();
+            }
+
+            return readAwaitable;
+        }
+
+        private void StopTimingRead(long bytesRead)
+        {
+            _context.TimeoutControl.BytesRead(bytesRead - _alreadyTimedBytes);
+            _alreadyTimedBytes = 0;
+
+            if (_backpressure)
+            {
+                _backpressure = false;
+                _context.TimeoutControl.StopTimingRead();
+            }
+        }
+
+        public override void Complete(Exception exception)
+        {
+            _context.ReportApplicationError(exception);
+            _completed = true;
+        }
+
+        public override void OnWriterCompleted(Action<Exception, object> callback, object state)
+        {
+            // TODO make this work with ContentLength.
+        }
+
+        public override void CancelPendingRead()
+        {
+            Interlocked.Exchange(ref _userCanceled, 1);
+            _context.Input.CancelPendingRead();
+        }
+
+        protected override Task OnStopAsync()
+        {
+            Complete(null);
+            return Task.CompletedTask;
+        }
+
+        protected override Task OnConsumeAsync()
```

Comment:
There's a lot of duplication in the OnConsumeAsync paths. Try to consolidate as much as possible. Virtual dispatch won't hurt perf too bad, and it could save us from making only a partial fix in the future.